backend/session/libseat.c

```diff
@@ -0,0 +1,211 @@
+	libseat_dispatch(seat, 0);
+	return 1;
+}
+
+static struct libseat_seat_listener seat_listener = {
+	.enable_seat = handle_enable_seat,
+	.disable_seat = handle_disable_seat,
+};
+
+static struct libseat_session *libseat_session_from_session(
+		struct wlr_session *base) {
+	assert(base->impl == &session_libseat);
+	return (struct libseat_session *)base;
+}
+
+static struct wlr_session *libseat_session_create(struct wl_display *disp) {
+	struct libseat_session *session = calloc(1, sizeof(*session));
+	if (!session) {
+		wlr_log(WLR_ERROR, "Allocation failed: %s", strerror(errno));
+		return NULL;
+	}
+
+	wl_list_init(&session->devices);
+
+	session->seat = libseat_open_seat(&seat_listener, session);
+	if (session->seat == NULL) {
+		wlr_log(WLR_ERROR, "Unable to create seat: %s\n", strerror(errno));
+		goto error;
+	}
+
+	const char *seat_name = libseat_seat_name(session->seat);
+	if (seat_name == NULL) {
+		wlr_log(WLR_ERROR, "Unable to get seat info: %s\n", strerror(errno));
+		goto error;
+	}
+	snprintf(session->base.seat, sizeof(session->base.seat), "%s", seat_name);
+
+	struct wl_event_loop *event_loop = wl_display_get_event_loop(disp);
+	session->event = wl_event_loop_add_fd(event_loop, libseat_get_fd(session->seat),
+		WL_EVENT_READABLE, libseat_event, session->seat);
+
+	if (libseat_dispatch(session->seat, 0) == -1 && errno != EAGAIN) {
+		wlr_log(WLR_ERROR, "libseat dispatch failed: %s\n", strerror(errno));
+		goto error;
+	}
+
+	if (!session->activated) {
+		wlr_log(WLR_INFO, "Waiting for seat activation on %s", session->base.seat);
+	}
+
+	while (!session->activated) {
+		if (wl_event_loop_dispatch(event_loop, -1) == -1) {
+			wlr_log(WLR_ERROR, "libseat dispatch failed: %s\n", strerror(errno));
+			goto error;
+		}
+	}
+
+	wlr_log(WLR_INFO, "Successfully loaded libseat session");
+	session->base.impl = &session_libseat;
+	session->signal = true;
+	return &session->base;
+
+error:
+	if (session->seat != NULL) {
+		libseat_close_seat(session->seat);
+		session->seat = NULL;
+	}
+	if (session->event != NULL) {
+		wl_event_source_remove(session->event);
+		session->event = NULL;
+	}
+	free(session);
+	return NULL;
+}
+
+static void libseat_session_destroy(struct wlr_session *base) {
+	struct libseat_session *session = libseat_session_from_session(base);
+
+	libseat_close_seat(session->seat);
+	wl_event_source_remove(session->event);
+	free(session);
+}
+
+static struct libseat_device *find_device_by_fd(struct libseat_session *session, int fd) {
+	struct libseat_device *dev;
+	wl_list_for_each(dev, &session->devices, link) {
+		if (dev->fd == fd) {
+			return dev;
+		}
+	}
+	return NULL;
+}
+
+static int libseat_session_open_device(struct wlr_session *base, const char *path) {
+	struct libseat_session *session = libseat_session_from_session(base);
+
+	int fd;
+	int device_id = libseat_open_device(session->seat, path, &fd);
+	if (device_id == -1) {
+		wlr_log(WLR_ERROR, "Failed to open device '%s': %s", path, strerror(errno));
```

Comment:
We have `wlr_log_errno` for this pattern

Reply:
Thanks!